testing/src/test/java/io/stargate/it/storage/StargateContainer.java

```diff
@@ -0,0 +1,528 @@
+      Integer.getInteger("stargate.test.process.wait.timeout.minutes", 10);
+
+  private static final AtomicBoolean suiteExecuting = new AtomicBoolean();
+  private static final AtomicBoolean testExecuting = new AtomicBoolean();
+
+  // the first 10 addresses are reserved for storage nodes
+  private static final AtomicInteger stargateAddressStart = new AtomicInteger(11);
+  private static final AtomicInteger stargateInstanceSeq = new AtomicInteger();
+
+  private static Container container;
+
+  static {
+    Thread containerShutdown =
+        new Thread("stargate-container-shutdown") {
+          @Override
+          public void run() {
+            try {
+              Container c = container;
+              if (c != null) {
+                c.stop();
+              }
+            } catch (Exception e) {
+              LOG.warn("Exception during global container shutdown: {}", e.getMessage(), e);
+            }
+          }
+        };
+
+    Runtime.getRuntime().addShutdownHook(containerShutdown);
+  }
+
+  private static Container container() {
+    if (container == null) {
+      throw new IllegalStateException("Stargate container has not been configured");
+    }
+
+    return container;
+  }
+
+  private static void enter(AtomicBoolean flag) {
+    if (!flag.compareAndSet(false, true)) {
+      throw new IllegalStateException(
+          String.format(
+              "Concurrent execution with %s is not supported",
+              StargateContainer.class.getSimpleName()));
+    }
+  }
+
+  private static void leave(AtomicBoolean flag) {
+    flag.compareAndSet(true, false);
+  }
+
+  private static StargateParameters parameters(StargateSpec spec, ExtensionContext context)
+      throws Exception {
+    Builder builder = StargateParameters.builder();
+
+    String customizer = spec.parametersCustomizer().trim();
+    if (!customizer.isEmpty()) {
+      Object testInstance = context.getTestInstance().orElse(null);
+      Class<?> testClass = context.getRequiredTestClass();
+      Method method = testClass.getMethod(customizer, Builder.class);
+      method.invoke(testInstance, builder);
+    }
+
+    return builder.build();
+  }
+
+  private static synchronized void start(ExtensionContext context) throws Exception {
+    ClusterConnectionInfo backend = ExternalStorage.connectionInfo();
+
+    StargateSpec spec =
+        context
+            .getElement()
+            .flatMap(e -> Optional.ofNullable(e.getAnnotation(StargateSpec.class)))
+            .orElseGet(
+                () ->
+                    context
+                        .getTestClass()
+                        .flatMap(c -> Optional.ofNullable(c.getAnnotation(StargateSpec.class)))
+                        .orElse(DefaultSpecHolder.get()));
+
+    StargateParameters params = parameters(spec, context);
+
+    Container c = container;
+    if (c != null) {
+      if (!c.matches(backend, spec, params)) {
+        c.stop();
+      } else {
+        // reuse container
```

Comment:
Maybe we could add a debug log to be able to trace this (if there is a need)

Reply:
Sure. will do. I think I'll and an INFO message for visibility.

Reply:
Added `LOG.info` in 3d0f45473fa69678a78113cf2d4c8d0f49753329